aten/src/ATen/cuda/CachingManagedAllocator.cpp

```diff
@@ -0,0 +1,1270 @@
+}
+
+void* raw_alloc(size_t nbytes) {
+  if (nbytes == 0) {
+    return nullptr;
+  }
+  int device;
+  C10_CUDA_CHECK(cudaGetDevice(&device));
+  void* r = nullptr;
+  allocator.malloc(
+      &r, device, nbytes, cuda::getCurrentCUDAStream(device));
+  return r;
+}
+
+void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream) {
+  if (nbytes == 0) {
+    return nullptr;
+  }
+  int device;
+  C10_CUDA_CHECK(cudaGetDevice(&device));
+  void* r = nullptr;
+  allocator.malloc(&r, device, nbytes, stream);
+  return r;
+}
+
+void raw_delete(void* ptr) {
+  allocator.free(ptr);
+}
+
+cudaError_t recordEvent(void *ptr, at::cuda::CUDAStream stream)
+{
+  return allocator.recordEvent(ptr, stream);
+}
+
+void emptyCache()
+{
+  allocator.emptyCache();
+}
+
+static void CachingManagedDeleter(void* ptr) {
+  allocator.free(ptr);
+}
+bool forceUncachedAllocator() {
+  static bool force_uncached =
+      getenv("PYTORCH_NO_CUDA_MEMORY_CACHING") != nullptr;
+  return force_uncached;
+}
+
+static void uncached_delete(void* ptr) {
+  C10_CUDA_CHECK(cudaFree(ptr));
+}
+
+struct CachingManagedAllocator final : public at::Allocator {
+  at::DataPtr allocate(size_t size) const override {
+    int device;
+    C10_CUDA_CHECK(cudaGetDevice(&device));
+    void* ptr = nullptr;
+    if (forceUncachedAllocator()) {
+      // Deliberately don't use cudaMallocMaybeCapturing here, to force an error
+      // if someone tries to use forceUncachedAllocator while capturing.
+      C10_CUDA_CHECK(cudaMallocManaged(&ptr, size));
+      return {ptr, ptr, &uncached_delete, Device(DeviceType::CUDA, device)};
+    }
+    if (size != 0) {
+      ptr = allocator.malloc(device, size, cuda::getCurrentCUDAStream(device))->ptr;
+      //C10_CUDA_CHECK(cudaMemAdvise(ptr, size, cudaMemAdviseSetPreferredLocation, device));
+    }
+    return {ptr, ptr, &CachingManagedDeleter, Device(DeviceType::CUDA, device)};
+  }
+  at::DeleterFnPtr raw_deleter() const override {
+    if (forceUncachedAllocator()) {
+      return &uncached_delete;
+    } else {
+      return &CachingManagedDeleter;
+    }
+  }
+};
+
+struct CachingManagedAllocatorCpu final : public at::Allocator {
+  at::DataPtr allocate(size_t size) const override {
+    void* ptr = nullptr;
+    if (forceUncachedAllocator()) {
+      // Deliberately don't use cudaMallocMaybeCapturing here, to force an error
+      // if someone tries to use forceUncachedAllocator while capturing.
+      C10_CUDA_CHECK(cudaMallocManaged(&ptr, size));
+      return {ptr, ptr, &uncached_delete, Device(DeviceType::CPU)};
+    }
+    if (size != 0) {
+      int device;
+      C10_CUDA_CHECK(cudaGetDevice(&device));
+      int hint_device = -1; //cudaCpuDeviceId
+      ptr = allocator.malloc(hint_device, size, cuda::getCurrentCUDAStream(device))->ptr;
+
+      //C10_CUDA_CHECK(cudaMemAdvise(ptr, size, cudaMemAdviseSetPreferredLocation, hint_device));
+      //C10_CUDA_CHECK(cudaMemAdvise(ptr, size, cudaMemAdviseSetAccessedBy, device));
```

Comment:
Should it be the default behavior to set the preferred location to CPU and accessed device to be the current device?

Reply:
Thanks for your interest! We debated a bit on this and this is an area that could certainly evolve with more testing/tuning.